tests/test_display_expression.cpp

```diff
@@ -0,0 +1,69 @@
+#include "../src/combinator.hpp"
+#include <iostream>
+#include <metricq/json.hpp>
+
+int main()
+{
+    try
+    {
+        // Input test data
+        nlohmann::json data = {
+            { "testCases",
+              { { { "expression",
+                    { { "operation", "*" },
+                      { "left", 5 },
+                      { "right", { { "operation", "-" }, { "left", 45 }, { "right", 3 } } } } } },
+                { { "expression",
+                    { { "operation", "*" },
+                      { "left", { { "operation", "+" }, { "left", 1 }, { "right", 2 } } },
+                      { "right",
+                        { { "operation", "-" },
+                          { "left", 10 },
+                          { "right", "dummy.source" } } } } } },
+                { { "expression",
+                    { { "operation", "-" },
+                      { "left",
+                        { { "operation", "+" },
+                          { "left", 15.3 },
+                          { "right",
+                            { { "operation", "min" }, { "inputs", { 42, 24, 8, 12 } } } } } },
+                      { "right",
+                        { { "operation", "throttle" },
+                          { "cooldown_period", "42" },
+                          { "input", 8 } } } } } } } }
+        };
+
+        // Expected outputs
+        std::vector<std::string> expected = { "(5 * (45 - 3))", "((1 + 2) * (10 - dummy.source))",
+                                              "((15.300000 + min[42, 24, 8, 12]) - 8)" };
```

Comment:
The formatting makes it really hard to see that we have three test cases with three expected strings. Maybe clang-format sees that differently, but I would appreciate a line break on line 37.

Reply:
Maybe it would also be better to couple the test case input more with the expected output.
Something simple like

```c++
struct TestCase {
  nlohmann::json input;
  std::string expected;
};
```

and then putting these things one by one into a vector with emplace_back.